solver/src/solver.rs

```diff
@@ -33,26 +38,151 @@ mod solver_utils;
 // minus this duration to account for additional delay for example from the network.
 const TIMEOUT_SAFETY_BUFFER: Duration = Duration::from_secs(5);
 
-/// Interface that all solvers must implement
+/// Interface that all solvers must implement.
+///
 /// A `solve` method transforming a collection of `Liquidity` (sources) into a list of
 /// independent `Settlements`. Solvers are free to choose which types `Liquidity` they
-/// would like to include/process (i.e. those already supported here or their own private sources)
-/// The `name` method is included for logging purposes.
+/// would like to process, including their own private sources.
 #[async_trait::async_trait]
 pub trait Solver {
     /// The returned settlements should be independent (for example not reusing the same user
     /// order) so that they can be merged by the driver at its leisure.
     async fn solve(&self, orders: Vec<Liquidity>, gas_price: f64) -> Result<Vec<Settlement>>;
 
-    /// Solver's account that should be used to submit settlements.
+    /// Returns solver's account that should be used to submit settlements.
     fn account(&self) -> &Account;
 
-    /// Displayable name of the solver. Defaults to the type name.
+    /// Returns displayable name of the solver.
+    ///
+    /// This method is used in logging and metrics. By default, returns the type name.
     fn name(&self) -> &'static str {
```

Comment:
I think it probably makes sense to get to the extra information like name and Address through the new Solvers struct. Would we remove these methods in the future then and store Address, Name in the Solvers struct?

Reply:
I'm not sure. These methods are just convenience shortcuts. Name and address is a property of a solver, why would we store them somewhere else.

In fact, I'm not sure that `Solvers` struct is the best approach. Having a reference to the solver instead of solver ID is more convenient. Another thing to take into account is this. If we configure solvers through config, then we may have multiple HTTP solvers with different names and URLs.

Reply:
We already have different HTTP solvers btw. Mip and Quasimodo.
If we do keep the information in the trait then I'm not sure if the Solvers struct is that useful. We could store solvers as Arc<dyn Solver> and directly use (Settlement, Arc< dyn Solver>) instead of the SolverId. Then we have all the information and less duplicated code.

Reply:
`Arc` is one way to go about it. Initially I didn't want to use `Arc` because we clone settlements here and there, so a lightweight identifier seemed preferable. But with current code structure (i.e. after `SettlementWithSolver` was replaced with a tuple), we can just use `&dyn Solver` everywhere.

Reply:
Maybe we can keep the SolverId struct as a value type that can be easily copiable for passing along with the settlement but make the solver expose a `solver_id` method that returns it. My worry about `&dyn Solver` would be that lifetimes may become challenging when we pass it around, but if this is not an issue then we can also do that.

Reply:
I'd also be OK with passing `Arc<dyn Solver>`s or `&dyn Solver`s around, then we can just read `account` and `name` from the solver directly.

> My worry about &dyn Solver would be that lifetimes may become challenging when we pass it around

AFAIU their lifetime is tied to `&self` that is passed in and also lives for the entire "run loop" future - so I'd imagine it wouldn't be too crazy.

That being said, the `SolverId(usize)` pattern is cool :smile:.